# NumPy

## Numpy의 소개



NumPy(Numerical Python)는 파이썬에서 과학적 계산을 위한 핵심 라이브러리이다. NumPy는 다차원 배열 객체와 배열과 함께 작동하는 도구들을 제공한다. 하지만 NumPy 자체로는 고수준의 데이터 분석 기능을 제공하지 않기 때문에 NumPy 배열과 배열 기반 컴퓨팅의 이해를 통해 pandas와 같은 도구를 좀 더 효율적으로 사용하는 것이 필요하다.

### ndarray 생성
array 함수를 사용해서 배열을 생성

In [2]:
import numpy as np

## ndarray 생성
arr = np.array([1,2,3,4])
print(arr)

[1 2 3 4]


In [3]:
np.zeros((3,3))

array([[0., 0., 0.],
       [0., 0., 0.],
       [0., 0., 0.]])

In [4]:
np.ones((2,2))

array([[1., 1.],
       [1., 1.]])

In [5]:
np.empty((4,4))

array([[4.66902623e-310, 1.81176639e-305, 1.33360301e+241,
        8.39037136e-311],
       [1.85505085e-229, 5.54041819e+228, 8.86790448e-310,
        9.11390253e-304],
       [1.27612992e-303, 1.27612715e-303, 8.20430637e-304,
        5.45857016e-217],
       [1.22164517e-173, 6.89528001e-246, 6.33289363e-140,
        7.45673900e-304]])

In [6]:
np.arange(10)

array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

In [9]:
np.arange(-5, 5, 1)

array([-5, -4, -3, -2, -1,  0,  1,  2,  3,  4])

## ndarray 배열의 모양, 차수, 데이터 타입을 확인하는 방법

In [10]:
arr = np.array([[1,2,3], [4,5,6]])

In [11]:
print(arr)

[[1 2 3]
 [4 5 6]]


In [12]:
arr.shape

(2, 3)

In [13]:
arr.ndim

2

In [14]:
arr.dtype

dtype('int64')

## ndarray 배열의 타입 변환
astype 함수를 사용해서 정수형 데이터 배열을 실수형 데이터 배열로 변환하기

In [16]:
arr_int = np.array([1,2,3,4])

In [17]:
arr_int.dtype

dtype('int64')

In [18]:
arr_float = arr_int.astype(np.float64)

In [19]:
arr_float.dtype

dtype('float64')

In [20]:
print(arr_float)

[1. 2. 3. 4.]


## 문자형 데이터 배열을 정수형 데이터 배열로 변환하기

In [21]:
arr_str = np.array(['1', '2', '3'])

In [22]:
arr_str.dtype

dtype('<U1')

In [23]:
arr_int  = arr_str.astype(np.int64)

In [24]:
arr_int.dtype

dtype('int64')

![Built-in Python types, ndarray data types](images/numpy-1.png
)

![Data Types](images/numpy-2.png)

## ndarray 배열의 연산
기본 연산자와 함수를 통해 배열 연산하기

In [28]:
arr1 = np.array([[1,2], [3,4]])
arr2 = np.array([[5,6], [7,8]])

In [29]:
arr1 + arr2

array([[ 6,  8],
       [10, 12]])

In [30]:
np.add(arr1, arr2)

array([[ 6,  8],
       [10, 12]])

In [31]:
arr1 * arr2

array([[ 5, 12],
       [21, 32]])

## numpy 배열의 연산은 연산자(+, -, *, /)나 함수(add, subtract, multiply, divide)로 가능

In [32]:
arr1.dot(arr2)

array([[19, 22],
       [43, 50]])

In [33]:
np.dot(arr1, arr2)

array([[19, 22],
       [43, 50]])

array 배열에 기본 연산자 *를 통한 곱은 행렬의 곱이 아닌 배열의 요소의 곱이므로, 행렬의 곱을 수행하기 위해서 dot 함수를 사용해야 한다.

## ndarray 배열 슬라이싱

In [36]:
arr = np.array([[1,2,3],[4,5,6],[7,8,9]])

In [37]:
arr_1 = arr[:2, 1:3]

In [38]:
print(arr_1)

[[2 3]
 [5 6]]


In [39]:
arr_1[0,0] = 5

In [40]:
arr_1

array([[5, 3],
       [5, 6]])

In [41]:
arr

array([[1, 5, 3],
       [4, 5, 6],
       [7, 8, 9]])

In [42]:
arr = np.array([[1,2,3], [4,5,6]])

In [43]:
idx = arr > 3

In [44]:
idx

array([[False, False, False],
       [ True,  True,  True]])

In [45]:
print(arr[idx])

[4 5 6]


idx의 값들은 arr의 값이 3보다 큰 것인지(True), 같거나 작은 것인지(False) 나타낸다.

### Wine Quality 데이터 설명



Wine Quality 데이터에는 포르투갈의 레드 와인과 화이트 와인 두 개의 데이터 셋이 있다. 각각의 데이터 셋 안에는 물리화학적 실험에 대한 12개의 변수들이 있다. 변수에 대한 설명은 다음과 같다.



![Wine Quality 데이터 변수 설명](images/numpy-3.png)

In [1]:
import numpy as np

redwine = np.loadtxt(fname = 'samples/winequality-red.csv', delimiter = ';', skiprows = 1)

In [2]:
print(redwine)

[[ 7.4    0.7    0.    ...  0.56   9.4    5.   ]
 [ 7.8    0.88   0.    ...  0.68   9.8    5.   ]
 [ 7.8    0.76   0.04  ...  0.65   9.8    5.   ]
 ...
 [ 6.3    0.51   0.13  ...  0.75  11.     6.   ]
 [ 5.9    0.645  0.12  ...  0.71  10.2    5.   ]
 [ 6.     0.31   0.47  ...  0.66  11.     6.   ]]


In [3]:
redwine.shape

(1599, 12)

### Wine Quality 데이터를 활용한 기초 통계 분석



NumPy에는 기본적인 배열 통계 메서드가 있다. sum, mean과 같은 기본적인 통계 메서드를 통해 3.2에서 불러온 redwine 데이터의 기초적인 통계 분석을 해본다. 실습할 통계 메서드는 다음과 같다.

![배열 통계 메서드](images/numpy-4.png)

In [4]:
print(redwine.sum())

152084.78194


In [5]:
print(redwine.mean())

7.926036165311652


In [9]:
print(redwine.mean(axis = 0))

[ 8.31963727  0.52782051  0.27097561  2.5388055   0.08746654 15.87492183
 46.46779237  0.99674668  3.3111132   0.65814884 10.42298311  5.63602251]


In [8]:
redwine[:,0].mean()

8.31963727329581

redwine.sum(): redwine 데이터 셋에 있는 모든 데이터의 합 <br>
redwine.mean(): redwine 데이터 셋에 있는 모든 데이터의 평균 <br>
redwine.mean(axis=0): redwine 데이터 셋의 열별(axis=0은 열을 뜻한다.) 평균, 이를 통해 각 변수별 평균을 알 수 있다. Fixed acidity의 평균은 8.31963722 이며, volatile acidity의 평균은 0.52782051 이다. <br>
redwine[:,0].mean(): redwine 데이터 셋의 첫번째 열 (0번째 열)의 모든 값들의 평균을 뜻한다. 즉, fixed acidity의 평균만 알 수 있다.

In [10]:
print(redwine.max(axis = 0))

[ 15.9       1.58      1.       15.5       0.611    72.      289.
   1.00369   4.01      2.       14.9       8.     ]


In [11]:
print(redwine.min(axis = 0))

[4.6     0.12    0.      0.9     0.012   1.      6.      0.99007 2.74
 0.33    8.4     3.     ]
